File: javascript/kindergarten-garden/kindergarten-garden.js
var Garden = function(garden, students) {
    if (students === undefined) {
        students = [
            "Alice", "Bob", "Charlie", "David",
            "Eve", "Fred", "Ginny", "Harriet",
            "Ileana", "Joseph", "Kincaid", "Larry"
        ];
    }
    students.sort();
    students = students.map(student => student.toLowerCase());
    let mapping = {
        "V": "violets",
        "R": "radishes",
        "C": "clover",
        "G": "grass",
    };

    // Garden Parsing
    let [top, bottom] = garden.split("\n");
    top     = top   .split('').map(letter => mapping[letter]);
    bottom  = bottom.split('').map(letter => mapping[letter]);

    garden = {};
    students.forEach((student, index) => {
        garden[student] = [];
        garden[student].push(top[index * 2], top[index * 2 + 1]);
        garden[student].push(bottom[index * 2], bottom[index * 2 + 1]);
    });
    return garden;
}

module.exports = Garden;

File: java/kindergarten-garden/src/main/java/KindergartenGarden.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

class KindergartenGarden {

    private HashMap<String, List<Plant>> garden;
    KindergartenGarden(String garden, String[] students) {
        String[] rows = garden.split("\n");
        String upperRow = rows[0];
        String lowerRow = rows[1];

        this.garden = new HashMap<String, List<Plant>>();
        List<Plant> plants;

        Arrays.sort(students);
        
        for (int i = 0; i < rows[0].length(); i += 2) {
            plants = Arrays.asList(
                Plant.getPlant(upperRow.charAt(i)),
                Plant.getPlant(upperRow.charAt(i + 1)),
                Plant.getPlant(lowerRow.charAt(i)),
                Plant.getPlant(lowerRow.charAt(i + 1))
            );
            this.garden.put(students[i / 2], plants);
        }
    }

    KindergartenGarden(String garden) {
        this(
            garden,
            new String[] {
                "Alice", "Bob", "Charlie", "David",
                "Eve", "Fred", "Ginny", "Harriet",
                "Ileana", "Joseph", "Kincaid", "Larry"
            });
    }

    List<Plant> getPlantsOfStudent(String student) {
        return this.garden.get(student);
    }
}